src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+                });
+
                 return executionId;
             }
-            catch
+            catch (Exception e)
             {
+                _logger.LogError(e, "An error occurred while setting up the tenant.");
+                context.Errors.Add(String.Empty, S["An error occurred while setting up the tenants: {0}", e.Message]);
+
                 context.ShellSettings.State = initialState;
-                await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                await _shellHost.ReloadShellContextAsync(shellSettings ?? context.ShellSettings, eventSource: false);
+
+                var shellScope = await _shellHost.GetScopeAsync(shellSettings ?? context.ShellSettings);
+                await shellScope.UsingAsync(async scope =>
+                {
+                    var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                    await tenandSetupHandlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
+                });
 
-                throw;
+                return null;
             }
         }
 
-        private async Task<string> SetupInternalAsync(SetupContext context)
+        private async Task<(string, ShellSettings)> SetupInternalAsync(SetupContext context)
         {
-            string executionId;
+            _logger.LogInformation("Running setup for tenant '{TenantName}'.", context.ShellSettings?.Name);
 
-            if (_logger.IsEnabled(LogLevel.Information))
-            {
-                _logger.LogInformation("Running setup for tenant '{TenantName}'.", context.ShellSettings.Name);
-            }
-
-            // Features to enable for Setup
-            string[] hardcoded =
+            // Features to enable for Setup.
+            string[] coreFeature =
             {
                 _applicationName,
                 "OrchardCore.Features",
                 "OrchardCore.Scripting",
                 "OrchardCore.Recipes"
             };
 
-            context.EnabledFeatures = hardcoded.Union(context.EnabledFeatures ?? Enumerable.Empty<string>()).Distinct().ToList();
-
-            // Set shell state to "Initializing" so that subsequent HTTP requests are responded to with "Service Unavailable" while Orchard is setting up.
-            context.ShellSettings.AsInitializing();
+            context.EnabledFeatures = coreFeature.Union(context.EnabledFeatures ?? Enumerable.Empty<string>()).Distinct().ToList();
 
             // Due to database collation we normalize the userId to lower invariant.
             // During setup there are no users so we do not need to check unicity.
-            context.Properties[SetupConstants.AdminUserId] = _setupUserIdGenerator.GenerateUniqueId().ToLowerInvariant();
+            context.Properties[SetupConstants.AdminUserId] = _setupUserIdGenerator.GenerateUniqueId();
```

Comment:
Hmm, maybe possible that we resolve a generator that doesn't only provide lower cases.

Reply:
Because of this comment

            // Due to database collation we normalize the userId to lower invariant.

Reply:
yes but wouldn't that be a problem everywhere since we use the same `GenerateUniqueId()` everywhere? If the collation can't support uppercases, then either the collation should be changed or the `GenerateUniqueId()`. changing it here only seems very bad to me.